woocommerce/woocommerce
# Plugin installer version 2

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR introduces a new version of plugin installer, allowing to install products purchased on WooCommerce.com much faster and without dependency on WP cron being available on the remote site. 



### How to test the changes in this Pull Request:



## Preconditions
- Ensure you have an active subscription to a product on WooCommerce.com. Jot down the product ID.
- Connect the store to your account on WooCommerce.com from **WP admin -> WooCommerce -> Extensions -> My subscriptions.**. Ensure you perform the connection from behalf of admin user of the store.
- Comment out [lines 102-110](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L102-L110) and [114-122](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L114-L122) in file `/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php` to bypass validation of authentication credentials and allow testing API endpoints through Curl.

## Test a successful installation
1. Send an installation request (ensure to replace strings `<SITE_DOMAIN>` and `<PRODUCT_ID>` with your site domain and product ID you got previously):
```bash
curl --request POST \
  --url 'https://<SITE_DOMAIN>/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY' \
  --header 'Content-Type: application/json' \
  --data '{
	"product-id": <PRODUCT_ID>,
	"run-until-step": "activate_product",
	"idempotency-key": "ABC"
}'
```
2. Wait for a few seconds and confirm you received a response on successful installation, with 200 code and response body similar to the following:
```json
{
	"success": true,
	"state": {
		"product_id": 473617,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Social Login",
		"already_installed_plugin_info": null,
		"started_seconds_ago": 11
	}
}
```
3. Go to site admin and confirm the plugin has been successfully installed and activated.

## Test an installation error
1. Retry the same request from the previous section.
2. Confirm you received an error response, with 400 code and response body stating the installation has already been run:
```json
{
	"success": false,
	"error_code": "all_installation_steps_run",
	"error_message": "All installation steps have been run",
	"state": {
		"product_id": 473617,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Social Login",
		"already_installed_plugin_info": {
			"name": "WooCommerce Social Login",
			"version": "2.12.0",
			"active": true
		},
		"started_seconds_ago": 365
	}
}
```

## Test installation state reset
1. Send a request to reset the installation state (ensure to replace strings `<SITE_DOMAIN>` and `<PRODUCT_ID>` with your site domain and product ID you got previously):
```bash
curl --request DELETE \
  --url 'https://<SITE_DOMAIN>/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY' \
  --header 'Content-Type: application/json' \
  --data '{
	"product-id": <PRODUCT_ID>,
	"idempotency-key": "ABC"
}'
```
2. Wait for a few seconds and confirm you received a successful response, with 200 code and response body similar to the following:
```json
{
	"success": true,
	"state": null
}
```
